_posts/2020-12-31-echo-supporting-old-app-versions.markdown

```diff
@@ -0,0 +1,66 @@
+
+```objc
+NSURL *url = [[NSURL alloc] initWithString:@"https://echo-api-production.herokuapp.com/"];
+```
+
+For whatever reason, we never created an Artsy-controlled CNAME DNS record for Echo. We were just hitting the bare Heroku URL directly. That means that there are _seven years'_ worth of Artsy app versions out there that _need_ to be able to continue hitting _that specific_ Heroku URL.
+
+😬
+
+This is what I meant earlier about how you need to think about properly-working code differently when you develop software that runs on someone else's hardware. If this was a server, this whole problem of old-code-hitting-outdated-APIs wouldn't exist. But because we ship apps as binaries that get ran on someone _else's_ hardware, we need to be aware of this kind of issue.
+
+That URL was fine when the code was written and it worked as intended for years. No one could have predicted, when it was added, that it would cause us headaches much later. Nor could we have anticipated that the Echo server's codebase would end up in such a state.
+
+## The Solution
+
+This is the part of the blog post where I get to be a hero (at least in my head). The Heroku app running Echo was on Cedar-14 and while we couldn't make any changes to its code, we _could_ replace the code entirely. This would give us a new server running at the old URL. So I wrote up a small Express server to proxy HEAD and GET API requests from old iOS app versions to return the response body and headers that were expected. [The work is here](https://github.com/artsy/echo/pull/59#) if you're curious. The nice thing about Heroku, at least, is that if this had gone horribly wrong we still could have reverted back to the Cedar-14 app while we figured out our next steps.
+
+We will still need to keep this Heroku app running, indefinitely, which isn't ideal. Perhaps Artsy's migration to Kubernetes will never be _quite_ 100% complete, but that's a small price to pay for keeping users of our app happy.
+
+## Conclusion
+
+It's not just bugs that you need to be aware of _accidentally_ shipping in your app binaries. You have to think about how the code that's running _as intended today_, because it will continue to run for the foreseeable future. This means adding checks for non-200 response codes from APIs, being careful about third-party APIs, and thinking carefully about everything you ship. Apps are more than just their code; apps are everything that their code depends on, too.
+
+This is a lesson that I've been teaching engineers for a long time, but it's only with our recent Echo changes that I've come to understand, at a deeper level, what it means to take ownership of code. Yes, the bugs, and of course, the happy little accidents, but also the mundane interconnected dependencies that make software systems so complex. And, if I'm being, that make software systems so much fun to work on, too.
+
```

Comment:
I think you're missing a word here, maybe `frank` or `honest` ?

Reply:
Yup, good catch 👍 I've updated that in another commit.